File: javascript/medium/subarraySumEqualsK.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var subarraySum = function(nums, k) {
    const map = new Map();
    let sum = 0, count = 0;
    for(let num of nums) {
        sum += num;
        if(sum === k) {
            count++;
        }
        if(map.has(sum-k)) {
            count += map.get(sum-k);
        }
        map.set(sum, map.has(sum) ? map.get(sum)+1 : 1);
    }
    return count;
};

File: ruby/medium/subarray_sum_equals_k.rb
# @param {Integer[]} nums
# @param {Integer} k
# @return {Integer}
def subarray_sum(nums, k)
    sum, count = 0, 0
    map = {}
    nums.each do |num|
        sum += num
        count += 1 if (sum == k)
        count += map[sum-k] if (map.has_key?(sum-k))
        map[sum] = (map.has_key?(sum) ? map[sum]+1 : 1)
    end
    return count
end
